src/rebar3_elixir_compile_resource.erl

```diff
@@ -79,10 +79,14 @@ fetch({elixir, Name_, Vsn_}, CDN) ->
     Vsn  = rebar3_elixir_compile_util:to_binary(Vsn_),
     case filelib:is_dir(Dir) of
         false ->
-            Package = binary_to_list(<<Name/binary, "-", Vsn/binary, ".tar">>),
-            Url = string:join([CDN, Package], "/"),
-            case request(Url) of
-                {ok, Binary} ->
+            Config = #{
+              http_adapter => hex_http_httpc,
+              http_adapter_config => #{profile => default},
+              http_user_agent_fragment => <<"(httpc)">>,
+              repo_url => list_to_binary(CDN)
+            },
+            case hex_repo:get_tarball(Config, Name, Vsn) of
+                {ok, Binary, _} ->
                     {ok, Contents} = extract(Binary),
                     ok = erl_tar:extract({binary, Contents}, [{cwd, Dir}, compressed]);
```

Comment:
You can use `hex_tarball:extract` instead of doing it manually. Besides contents and metadata it returns tarball checksum which should be checked against registry to make sure it hasnt been tampered with.

Reply:
There's no specific function to check tarball against registry, you have to do it manually. You can see how we do it on Hex: https://github.com/hexpm/hex/blob/v0.18.2/lib/hex/scm.ex#L139:L142.

The checksum is usually stored somewhere, e.g. in mix.lock/rebar.lock, cache etc and ideally it would be read from there. If you don't have it, you can grab with this: (I'm pretty printing it for nicer output, we should compare bytes.)

```erlang
{ok, {200, _, Releases}} = hex_repo:get_package(hex_core:default_config(), <<"hex_core">>),
 [{Version, hex_tarball:format_checksum(Checksum)} || #{version := Version, checksum := Checksum} <- Releases].
[{<<"0.1.0">>,
  <<"71CFC9410A85AC9386A8526E572623C7856F8939633C23AE3F32A6B49D9B899F">>},
 {<<"0.1.1">>,
  <<"098CC084BB6E6307C7A7EDDB0D8FEE8A107AF4D64F0A9573F7A377396D09ACD2">>},
 {<<"0.2.0">>,
  <<"3A7EACCFB8ADD3FF05D950C10ED5BDB5D0C48C988EBBC5D7AE2A55498F0EFF1B">>},
 {<<"0.2.1">>,
  <<"AC4A05DF753A728866A5BC0A11CB9719684C2DCD1837CE96392061B01CC69258">>},
 {<<"0.3.0">>,
  <<"99492AE7111C4C7215C2FF4E62FF7C3A5AB7C5202824883372ECD2E97907A831">>}]
```

Reply:
Thanks! @benoitc what do you think - should this plugin use checksum stored in rebar.lock/mix.lock or can it take it from Hex?